The matplotlib lazy import itself works correctly — preflight runs end-to-end in a torch + pyyaml venv without --plot. 

Like this command:

```bash
srun -t 00:15:00 -N 1 -c 128 --gpus-per-node=8 \
  runner/primus-cli-direct-preflight.sh \
  -- preflight --perf-test --plot --disable-pdf --report-file-name preflight-neg-plot-1N \
  2>&1 | tee preflight-neg-plot-1N.log
```

⚠️However, when --plot is used without matplotlib installed (I didn't do `uv pip install requirements.txt` in my virtual env) , the GPU rank 0's `ModuleNotFoundError` inside `run_square_gemm`'s `if RANK == 0:` block causes a cross-rank deadlock: rank 0 drops into `finalize_distributed()`'s `dist.barrier()` while ranks 1-7 continue into the next benchmark. The observable failure is a 120s NCCL ALLREDUCE watchdog timeout + SIGABRT on rank 0, not the underlying `ModuleNotFoundError`. 

Log snippet: 
<details>

```bash
[rank0]:[E424 19:39:22.409643535 ProcessGroupNCCL.cpp:689] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=8, OpType=ALLREDUCE, NumelIn=1, NumelOut=1, Timeout(ms)=120000) ran for 120016 milliseconds before timing out.
[rank0]:[E424 19:39:22.409830141 ProcessGroupNCCL.cpp:2303] [PG ID 0 PG GUID 0(default_pg) Rank 0]  failure detected by watchdog at work sequence id: 8 PG status: last enqueued work: 8, last completed work: 7
[rank0]:[E424 19:39:24.035945261 ProcessGroupNCCL.cpp:733] Stack trace of the failed collective: 
#0 barrier from /mnt/shared/ama10002/envs/preflight/.venv/lib/python3.12/site-packages/torch/distributed/distributed_c10d.py:5030
#1 wrapper from /mnt/shared/ama10002/envs/preflight/.venv/lib/python3.12/site-packages/torch/distributed/c10d_logger.py:83
#2 finalize_distributed from /mnt/shared/ama10002/Primus/primus/tools/utils.py:57
#3 run_preflight from /mnt/shared/ama10002/Primus/primus/tools/preflight/preflight_perf_test.py:330
#4 run from /mnt/shared/ama10002/Primus/primus/cli/subcommands/preflight.py:32
```
</details> 

Suggested changes:

- Add one-line note in `docs/run-preflight-without-container.md` that --plot and probably the PDF options require full `pip install -r requirements.txt`
- Maybe later write code to do a Precheck for matplotlib library on all ranks at the start of run_preflight when `--plot` is used

Example:
In `run_preflight()` in `preflight_perf_test.py`
<details> 

```python
def run_preflight(args):
    """..."""
    perf_test = getattr(args, "perf_test", False)

    # Fail-fast on all ranks if an optional feature is requested without its
    # dependency installed. Otherwise rank 0's lazy import can crash mid-run
    # and deadlock with other ranks still executing collectives.
    if getattr(args, "plot", False):
        try:
            import matplotlib  # noqa: F401
        except ImportError as e:
            raise ImportError(
                "--plot requires matplotlib. Install it (e.g. `pip install matplotlib`) "
                "or re-run without --plot."
            ) from e
  ```

</details> 

For now I'll probably just make a note in the docs. (will push this in a later commit) 

It is similar for the pdf report option. The lazy import is appears correct but the error message is easier to see this time

Here's the command I tried for the reports:

<details> 

```bash
srun -t 00:15:00 -N 1 -c 128 --gpus-per-node=8 \
  runner/primus-cli-direct-preflight.sh \
  -- preflight --perf-test --report-file-name preflight-neg-pdf-1N \
  2>&1 | tee preflight-neg-pdf-1N.log
  ```
Error logs:
```bash
Traceback (most recent call last):
  File "/mnt/shared/ama10002/Primus/primus/cli/main.py", line 172, in main
    args.func(args, unknown_args)
  File "/mnt/shared/ama10002/Primus/primus/cli/subcommands/preflight.py", line 32, in run
    rc = run_preflight(args)
         ^^^^^^^^^^^^^^^^^^^
  File "/mnt/shared/ama10002/Primus/primus/tools/preflight/preflight_perf_test.py", line 322, in run_preflight
    md_to_pdf(args.markdown_file, args.pdf_file)
  File "/mnt/shared/ama10002/Primus/primus/tools/preflight/utility.py", line 71, in md_to_pdf
    import markdown2
ModuleNotFoundError: No module named 'markdown2'
[Primus] Error: ModuleNotFoundError: No module named 'markdown2' (/mnt/shared/ama10002/Primus/primus/tools/preflight/utility.py:71)
E0424 20:10:23.846000 601388 torch/distributed/elastic/multiprocessing/api.py:986] failed (exitcode: 1) local_rank: 0 (pid: 601456) of binary: /mnt/shared/ama10002/envs/preflight/.venv/bin/python3
Traceback (most recent call last):
  File "/mnt/shared/ama10002/envs/preflight/.venv/bin/torchrun", line 10, in <module>
    sys.exit(main())
```
  
</details>

--- a/docs/run-preflight-without-container.md
+++ b/docs/run-preflight-without-container.md
@@ -26,20 +26,23 @@ There is a list of package in the project's top-level `requirements.txt`.
 
 ### Create virtual environment and install packages
 
-The environment is shared between all nodes in the cluster. So we need to create a virtual environment and install packages on a shared file system. Python venv or uv is recommended to create a virtual environment and install packages. Here is an example of creating a virtual environment and installing packages using uv: 
+The environment is shared between all nodes in the cluster. So we need to create a virtual environment and install packages on a shared file system. Python venv or uv is recommended to create a virtual environment and install packages. Here is an example of creating a virtual environment and installing packages using uv:
+
+#### Basic installation
 
 ```bash
-# Use a shared file system accessible from all nodes (e.g. NFS home, /shared)
 mkdir -p ~/envs/preflight
 cd ~/envs/preflight
 
 uv venv --python 3.12
 source .venv/bin/activate
 
-# install torch for rocm
 uv pip install torch torchvision --index-url https://download.pytorch.org/whl/rocm7.1
+```
 
-# install other packages
+#### Full installation (for plotting and PDF reports)
+
+```bash
 cd /path/to/Primus
 uv pip install -r requirements.txt
 ```

--- a/primus/core/config/yaml_loader.py
+++ b/primus/core/config/yaml_loader.py
@@ -7,8 +7,6 @@
 import os
 import re
 
-import yaml
-
 from primus.core.config.merge_utils import deep_merge
 
 ENV_PATTERN = re.compile(r"\${([^:{}]+)(?::([^}]*))?}")
@@ -30,6 +28,8 @@ def parse_yaml(path: str) -> dict:
 # 1. Load YAML
 # ================================================================
 def _load_yaml(path: str):
+    import yaml
+
     with open(path, "r") as f:
         return yaml.load(f, Loader=yaml.SafeLoader)
 

--- a/primus/core/utils/yaml_utils.py
+++ b/primus/core/utils/yaml_utils.py
@@ -8,8 +8,6 @@ import json
 from types import SimpleNamespace
 from typing import Any, Mapping
 
-import yaml
-
 from primus.core.config.merge_utils import deep_merge
 from primus.core.config.yaml_loader import parse_yaml as _parse_yaml_core
 
@@ -140,6 +138,7 @@ def dump_namespace_to_yaml(ns: SimpleNamespace, file_path: str):
         >>> ns = SimpleNamespace(a=1, b=SimpleNamespace(c=2))
         >>> dump_namespace_to_yaml(ns, "config.yaml")
     """
+    import yaml
 
     def ns_to_dict(obj):
         if isinstance(obj, SimpleNamespace):

